File: Data Structures & Algorithms/string-encode-and-decode/submission-2.java
class Solution {

    public String encode(List<String> strs) {
        String encoded = "";
        int n = strs.size();
        for(String curr:strs){
            int len = curr.length();
            String lenS = Integer.toString((Integer)len);
            lenS += '#';
            encoded += lenS;
            encoded += curr;
        }
        return encoded;
    }

    public List<String> decode(String str) {
        List<String> ans = new ArrayList<>();
        int idx = 0;
        int n = str.length();
        while(idx < n){
            int len = 0;
            while(str.charAt(idx) != '#'){
                int dig = (str.charAt(idx)-'0');
                len = len*10 + dig;
                idx++;
            }
            idx++;
            String curr = "";
            for(int i = 0;i<len;i++){
                curr += str.charAt(idx);
                idx++;
            }
            ans.add(curr);
        }
        return ans;
    }
}

File: Data Structures & Algorithms/valid-parenthesis-string/submission-2.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int n = s.size();
        vector<int> nxt(n+2,0);
     
        nxt[0] = 1;
        for(int idx = n-1;idx>=0;idx--){
            vector<int> curr(n+2,0);
            for(int open = 0;open<n;open++){
                if(s[idx] == '('){
                    curr[open] |= nxt[open+1];
                }
                else if(s[idx] == ')'){
                    if(open-1 >= 0)
                        curr[open] = nxt[open-1];
                }
                else{
                    curr[open] = (nxt[open+1] | nxt[open]);
                    if(open-1 >= 0){
                        curr[open] |= nxt[open-1];
                    }
                }
            }
            nxt = curr;
            
        }
        // for(int i = 0;i<=n;i++){
        //     cout<<nxt[i]<<" ";
        // }
        return nxt[0];
    }
};
